### PR TITLE
Rspace/simplify main rspace algorithm

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
<sup>_What this PR does, and why it's needed_</sup>
As part of handoff of RSpace it makes sense to simplify the methods and streamline their usages.
RSpace and ReplayRSpace have a lot in common,  but it's obfuscated by implementation details.
This PRs goal is to make it obvious which parts are the same and which parts differ.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>
https://rchain.atlassian.net/browse/RCHAIN-3874
https://rchain.atlassian.net/browse/RCHAIN-3875

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by ArturGajowy** [comment]
@dzajkowski FTFY :)

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by kayvank** [comment]
try bors

**Comment by kayvank** [comment]
bors try

**Review by kayvank — APPROVED** [review]

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 1]
```diff
@@ -223,107 +222,84 @@ class RSpace[F[_], C, P, A, K](
...
-          _ <- logF.debug(
-                s"produce: matching continuation found at <channels: $channels>"
-              )
-        } yield constructResult
-    }
+    for {
+      _               <- metricsF.incrementCounter(produceCommLabel)
+      _               <- logComm(consumeRef, peeks, dataCandidates)
+      _               <- store.removeContinuation(channels, continuationIndex).unlessA(persistK)
+      indexedChannels = channels.zipWithIndex.toMap
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: Here we're creating an intermediate collection (via `zipWithIndex`) for no good reason. It's generally better to use `.iterator` in cases like this

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 2]
```diff
@@ -223,107 +222,84 @@ class RSpace[F[_], C, P, A, K](
...
+        peeks,
+        consumeRef
+      ),
+      continuationIndex,
+      dataCandidates
+    ) = pc
+
+    def removeMatchedDatumAndJoin(channelsToIndex: Map[C, Int]): F[Seq[Unit]] =
+      dataCandidates
+        .sortBy(_.datumIndex)(Ordering[Int].reverse)
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: Why does it have to be sorted?

Comment (marcin-rzeznicki) [reply rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: Looks like we're traversing it wholly anyway and we're interested only in `dataIndex >= 0`

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 3]
```diff
@@ -52,6 +52,100 @@ class RSpace[F[_], C, P, A, K](
...
 
+  override def consume(
+      channels: Seq[C],
+      patterns: Seq[P],
+      continuation: K,
+      persist: Boolean,
+      peeks: SortedSet[Int] = SortedSet.empty
+  ): F[MaybeActionResult] =
+    contextShift.evalOn(scheduler) {
+      if (channels.isEmpty) {
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: Yeah, right - so use a data structure that does not permit emptiness - also in the spirit of "making illegal states unrepresentable", one should introduce a structure that enforces that channels and patterns are of the same length (eg. a non-empty sequence of pairs)

File: rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala [hunk 4]
```diff
@@ -38,6 +40,14 @@ abstract class RSpaceOps[F[_]: Concurrent: Metrics, C, P, A, K](
         .fold(underlying.+((elem, 1)))(currentCount => underlying.+((elem, currentCount + 1)))
   }
 
+  implicit class RichFSeq[M[_]: Applicative, D](t: M[Seq[D]]) {
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala:50]: Very nice

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala [hunk 5]
```diff
@@ -223,29 +99,123 @@ class ReplayRSpace[F[_]: Sync, C, P, A, K](
...
+          } yield wrapResult(channels, patterns, continuation, persist, peeks, consumeRef, mats)
+    } yield r
+
+// TODO: refactor this monster
+  def getCommAndDataCandidates(
+      channels: Seq[C],
+      patterns: Seq[P],
+      comms: Seq[COMM]
+  ): F[Option[(COMM, Seq[DataCandidate[C, A]])]] = {
+    type COMMOrData = Either[COMM, (COMM, Seq[DataCandidate[C, A]])]
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: Myabe an ADT would be more appropriate for this?

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala [hunk 6]
```diff
@@ -410,27 +238,188 @@ class ReplayRSpace[F[_]: Sync, C, P, A, K](
...
+      data: A,
+      persist: Boolean,
+      comms: Seq[COMM],
+      produceRef: Produce,
+      groupedChannels: Seq[Seq[C]]
+  ): F[Either[COMM, (COMM, ProduceCandidate[C, P, A, K])]] = {
+    type COMMOrProduce = Either[COMM, (COMM, ProduceCandidate[C, P, A, K])]
+    def go(comms: Seq[COMM]): F[Either[Seq[COMM], COMMOrProduce]] =
+      comms match {
+        case Nil =>
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: Again, why not use `NonEmptyList[COMM]`

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 7]
```diff
@@ -204,54 +169,36 @@ class RSpace[F[_], C, P, A, K](
...
-          } yield firstMatch match {
-            case None             => remaining.asLeft[MaybeProduceCandidate]
-            case produceCandidate => produceCandidate.asRight[Seq[CandidateChannels]]
+  ): F[MaybeProduceCandidate] =
+    runMatcherForChannels(
+      groupedChannels,
+      channels =>
+        store
+          .getContinuations(channels)
+          .shuffleWithIndex,
```
Comment (marcin-rzeznicki) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala:143]: Why do we need to shuffle the whole collection here?

**Comment by marcin-rzeznicki** [comment]
@kayvank I've had a few comments, but I appreciate the initiative and think this is a change for the better. If you are OK with it, ping me, I'll aprove

**Review by tgrospic — APPROVED** [review]
With @marcin-rzeznicki comments LGTM

**Pull request merged**